WebDevStudios/wd_s
# Spike/mle abstract tailwind imports

Experimental Solution for: #614 

### DESCRIPTION

Even though we're likely going a different direction, this spike branch does the following:

1. Abstract Tailwind into a separate webpack entry.
2. Creates `style.scss` and `editor.scss` for importing styles
3. Update enqueues similar to #600 since these updates are very tightly coupled with this update
4. Add `@tailwind base;` to `style.scss` and `editor.scss` so that we can control the behavior behind this feature and address some concerns around selector name scoping.

See `spike/mle-abstract-tailwind-imports`

### Things not setup in this branch:
- Selector namespace updates are not part of this PR. This was part of the discussion but is a different issue. Importing `@tailwind base;` will make this issue easier to resolve and iterate on.
- CSS Mapping in the editor `DevTools failed to load SourceMap: Could not load content for http://wdsunderscores.test/wp-admin/index.css.map: HTTP error: status code 404, net::ERR_HTTP_RESPONSE_CODE_FAILURE` 

**Results**
_Times will vary depending on your machine._

- Rebuild times appear to range from `3500ms-4000ms` on my local using my MBP. Which is not much more than before these changes. Times will vary depending on what was change for the rebuild. On my Mac M1, rebuild times are around `2000ms-2300ms`.
- Regular build times appear to be around the same time.

### SCREENSHOTS
Examples below are taken from my Mac Mini M1. Results WILL vary according to you machine.

Editor styles being loaded with `@apply bg-purple-500;` on `.editor-post-title `.
<img width="847" alt="Screen Shot 2021-03-26 at 9 55 01 AM" src="https://user-images.githubusercontent.com/25035900/112642676-ed360880-8e19-11eb-9fd7-77e84f8f6fd4.png">

Rebuild example when adding style `@apply bg-purple-500;` on `.editor-post-title `.
<img width="1115" alt="Screen Shot 2021-03-26 at 9 54 53 AM" src="https://user-images.githubusercontent.com/25035900/112642677-edce9f00-8e19-11eb-8bba-1abb49b99a83.png">

Rebuild example when adding style `@apply bg-yellow-700;` on `body`.
<img width="1127" alt="Screen Shot 2021-03-26 at 9 53 49 AM" src="https://user-images.githubusercontent.com/25035900/112642680-ee673580-8e19-11eb-843f-75bd094c0973.png">

Initial build load time.
<img width="1078" alt="Screen Shot 2021-03-26 at 9 51 52 AM" src="https://user-images.githubusercontent.com/25035900/112642682-eeffcc00-8e19-11eb-9df1-14c966396686.png">


### OTHER

- [NA] Is this issue accessible? (Section 508/WCAG 2.0AA)
- [NA] Does this pass CBT?

### STEPS TO VERIFY

1. Checkout and build files.
2. Verify that styles and tailwind styles on the frontend work.
3. Verify that no consoles errors are triggered on the frontend.
4. Verify that styles and tailwind styles in the editor work.
5. Verify that no consoles errors are triggered in the editor.

### DOCUMENTATION

Will this pull request require updating the wd_s [wiki](https://github.com/WebDevStudios/wd_s/wiki)?

Not yet, as this feature is experimental.
